{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6ea646f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np, matplotlib.pyplot as plt\n",
    "import sys, os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a4eceb3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append(os.path.join(os.getcwd(), \"build\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9c7d9588",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting volpy\n",
      "  Using cached volpy-21.4.1-py3-none-any.whl.metadata (588 bytes)\n",
      "Collecting numpy==1.20.0 (from volpy)\n",
      "  Using cached numpy-1.20.0.zip (8.0 MB)\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'error'\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  × Preparing metadata (pyproject.toml) did not run successfully.\n",
      "  │ exit code: 1\n",
      "  ╰─> [271 lines of output]\n",
      "      setup.py:66: RuntimeWarning: NumPy 1.20.0 may not yet support Python 3.11.\n",
      "        warnings.warn(\n",
      "      Running from numpy source directory.\n",
      "      setup.py:485: UserWarning: Unrecognized setuptools command, proceeding with generating Cython sources and expanding templates\n",
      "        run_build = parse_setuppy_commands()\n",
      "      Cythonizing sources\n",
      "      Processing numpy/random\\_bounded_integers.pxd.in\n",
      "      Processing numpy/random\\bit_generator.pyx\n",
      "      Processing numpy/random\\mtrand.pyx\n",
      "      Processing numpy/random\\_bounded_integers.pyx.in\n",
      "      Processing numpy/random\\_common.pyx\n",
      "      Processing numpy/random\\_generator.pyx\n",
      "      Processing numpy/random\\_mt19937.pyx\n",
      "      Processing numpy/random\\_pcg64.pyx\n",
      "      Processing numpy/random\\_philox.pyx\n",
      "      Processing numpy/random\\_sfc64.pyx\n",
      "      blas_opt_info:\n",
      "      blas_mkl_info:\n",
      "      No module named 'numpy.distutils._msvccompiler' in numpy.distutils; trying from distutils\n",
      "      customize MSVCCompiler\n",
      "        libraries mkl_rt not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      blis_info:\n",
      "        libraries blis not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      openblas_info:\n",
      "        libraries openblas not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "      get_default_fcompiler: matching types: '['gnu', 'intelv', 'absoft', 'compaqv', 'intelev', 'gnu95', 'g95', 'intelvem', 'intelem', 'flang']'\n",
      "      customize GnuFCompiler\n",
      "      Could not locate executable g77\n",
      "      Could not locate executable f77\n",
      "      customize IntelVisualFCompiler\n",
      "      Could not locate executable ifort\n",
      "      Could not locate executable ifl\n",
      "      customize AbsoftFCompiler\n",
      "      Could not locate executable f90\n",
      "      customize CompaqVisualFCompiler\n",
      "      Could not locate executable DF\n",
      "      customize IntelItaniumVisualFCompiler\n",
      "      Could not locate executable efl\n",
      "      customize Gnu95FCompiler\n",
      "      Could not locate executable gfortran\n",
      "      Could not locate executable f95\n",
      "      customize G95FCompiler\n",
      "      Could not locate executable g95\n",
      "      customize IntelEM64VisualFCompiler\n",
      "      customize IntelEM64TFCompiler\n",
      "      Could not locate executable efort\n",
      "      Could not locate executable efc\n",
      "      customize PGroupFlangCompiler\n",
      "      Could not locate executable flang\n",
      "      don't know how to compile Fortran code on platform 'nt'\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      atlas_3_10_blas_threads_info:\n",
      "      Setting PTATLAS=ATLAS\n",
      "        libraries tatlas not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      atlas_3_10_blas_info:\n",
      "        libraries satlas not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      atlas_blas_threads_info:\n",
      "      Setting PTATLAS=ATLAS\n",
      "        libraries ptf77blas,ptcblas,atlas not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      atlas_blas_info:\n",
      "        libraries f77blas,cblas,atlas not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\system_info.py:1989: UserWarning:\n",
      "          Optimized (vendor) Blas libraries are not found.\n",
      "          Falls back to netlib Blas library which has worse performance.\n",
      "          A better performance should be easily gained by switching\n",
      "          Blas library.\n",
      "        if self._calc_info(blas):\n",
      "      blas_info:\n",
      "        libraries blas not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\system_info.py:1989: UserWarning:\n",
      "          Blas (http://www.netlib.org/blas/) libraries not found.\n",
      "          Directories to search for the libraries can be specified in the\n",
      "          numpy/distutils/site.cfg file (section [blas]) or by setting\n",
      "          the BLAS environment variable.\n",
      "        if self._calc_info(blas):\n",
      "      blas_src_info:\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\system_info.py:1989: UserWarning:\n",
      "          Blas (http://www.netlib.org/blas/) sources not found.\n",
      "          Directories to search for the sources can be specified in the\n",
      "          numpy/distutils/site.cfg file (section [blas_src]) or by setting\n",
      "          the BLAS_SRC environment variable.\n",
      "        if self._calc_info(blas):\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      non-existing path in 'numpy\\\\distutils': 'site.cfg'\n",
      "      lapack_opt_info:\n",
      "      lapack_mkl_info:\n",
      "        libraries mkl_rt not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      openblas_lapack_info:\n",
      "        libraries openblas not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      openblas_clapack_info:\n",
      "        libraries openblas,lapack not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      flame_info:\n",
      "        libraries flame not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      atlas_3_10_threads_info:\n",
      "      Setting PTATLAS=ATLAS\n",
      "        libraries lapack_atlas not found in c:\\Users\\raphe\\Documents\\GitHub\\C-vol-surface-calib-engine\\.venv\\lib\n",
      "        libraries tatlas,tatlas not found in c:\\Users\\raphe\\Documents\\GitHub\\C-vol-surface-calib-engine\\.venv\\lib\n",
      "        libraries lapack_atlas not found in C:\\\n",
      "        libraries tatlas,tatlas not found in C:\\\n",
      "      <class 'numpy.distutils.system_info.atlas_3_10_threads_info'>\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      atlas_3_10_info:\n",
      "        libraries lapack_atlas not found in c:\\Users\\raphe\\Documents\\GitHub\\C-vol-surface-calib-engine\\.venv\\lib\n",
      "        libraries satlas,satlas not found in c:\\Users\\raphe\\Documents\\GitHub\\C-vol-surface-calib-engine\\.venv\\lib\n",
      "        libraries lapack_atlas not found in C:\\\n",
      "        libraries satlas,satlas not found in C:\\\n",
      "      <class 'numpy.distutils.system_info.atlas_3_10_info'>\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      atlas_threads_info:\n",
      "      Setting PTATLAS=ATLAS\n",
      "        libraries lapack_atlas not found in c:\\Users\\raphe\\Documents\\GitHub\\C-vol-surface-calib-engine\\.venv\\lib\n",
      "        libraries ptf77blas,ptcblas,atlas not found in c:\\Users\\raphe\\Documents\\GitHub\\C-vol-surface-calib-engine\\.venv\\lib\n",
      "        libraries lapack_atlas not found in C:\\\n",
      "        libraries ptf77blas,ptcblas,atlas not found in C:\\\n",
      "      <class 'numpy.distutils.system_info.atlas_threads_info'>\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      atlas_info:\n",
      "        libraries lapack_atlas not found in c:\\Users\\raphe\\Documents\\GitHub\\C-vol-surface-calib-engine\\.venv\\lib\n",
      "        libraries f77blas,cblas,atlas not found in c:\\Users\\raphe\\Documents\\GitHub\\C-vol-surface-calib-engine\\.venv\\lib\n",
      "        libraries lapack_atlas not found in C:\\\n",
      "        libraries f77blas,cblas,atlas not found in C:\\\n",
      "      <class 'numpy.distutils.system_info.atlas_info'>\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      lapack_info:\n",
      "        libraries lapack not found in ['c:\\\\Users\\\\raphe\\\\Documents\\\\GitHub\\\\C-vol-surface-calib-engine\\\\.venv\\\\lib', 'C:\\\\']\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\system_info.py:1849: UserWarning:\n",
      "          Lapack (http://www.netlib.org/lapack/) libraries not found.\n",
      "          Directories to search for the libraries can be specified in the\n",
      "          numpy/distutils/site.cfg file (section [lapack]) or by setting\n",
      "          the LAPACK environment variable.\n",
      "        return getattr(self, '_calc_info_{}'.format(name))()\n",
      "      lapack_src_info:\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\system_info.py:1849: UserWarning:\n",
      "          Lapack (http://www.netlib.org/lapack/) sources not found.\n",
      "          Directories to search for the sources can be specified in the\n",
      "          numpy/distutils/site.cfg file (section [lapack_src]) or by setting\n",
      "          the LAPACK_SRC environment variable.\n",
      "        return getattr(self, '_calc_info_{}'.format(name))()\n",
      "        NOT AVAILABLE\n",
      "      \n",
      "      numpy_linalg_lapack_lite:\n",
      "        FOUND:\n",
      "          language = c\n",
      "          define_macros = [('HAVE_BLAS_ILP64', None), ('BLAS_SYMBOL_SUFFIX', '64_')]\n",
      "      \n",
      "      C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py:275: UserWarning: Unknown distribution option: 'define_macros'\n",
      "        warnings.warn(msg)\n",
      "      running dist_info\n",
      "      running build_src\n",
      "      build_src\n",
      "      building py_modules sources\n",
      "      creating build\n",
      "      creating build\\src.win-amd64-3.11\n",
      "      creating build\\src.win-amd64-3.11\\numpy\n",
      "      creating build\\src.win-amd64-3.11\\numpy\\distutils\n",
      "      building library \"npymath\" sources\n",
      "      Traceback (most recent call last):\n",
      "        File \"c:\\Users\\raphe\\Documents\\GitHub\\C-vol-surface-calib-engine\\.venv\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 389, in <module>\n",
      "          main()\n",
      "        File \"c:\\Users\\raphe\\Documents\\GitHub\\C-vol-surface-calib-engine\\.venv\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 373, in main\n",
      "          json_out[\"return_val\"] = hook(**hook_input[\"kwargs\"])\n",
      "                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"c:\\Users\\raphe\\Documents\\GitHub\\C-vol-surface-calib-engine\\.venv\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 175, in prepare_metadata_for_build_wheel\n",
      "          return hook(metadata_directory, config_settings)\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 157, in prepare_metadata_for_build_wheel\n",
      "          self.run_setup()\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 249, in run_setup\n",
      "          self).run_setup(setup_script=setup_script)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 142, in run_setup\n",
      "          exec(compile(code, __file__, 'exec'), locals())\n",
      "        File \"setup.py\", line 513, in <module>\n",
      "          setup_package()\n",
      "        File \"setup.py\", line 505, in setup_package\n",
      "          setup(**metadata)\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\core.py\", line 169, in setup\n",
      "          return old_setup(**new_attr)\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\__init__.py\", line 165, in setup\n",
      "          return distutils.core.setup(**attrs)\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\core.py\", line 148, in setup\n",
      "          dist.run_commands()\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py\", line 967, in run_commands\n",
      "          self.run_command(cmd)\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py\", line 986, in run_command\n",
      "          cmd_obj.run()\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\command\\dist_info.py\", line 31, in run\n",
      "          egg_info.run()\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\command\\egg_info.py\", line 24, in run\n",
      "          self.run_command(\"build_src\")\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\cmd.py\", line 313, in run_command\n",
      "          self.distribution.run_command(command)\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py\", line 986, in run_command\n",
      "          cmd_obj.run()\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\command\\build_src.py\", line 144, in run\n",
      "          self.build_sources()\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\command\\build_src.py\", line 155, in build_sources\n",
      "          self.build_library_sources(*libname_info)\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\command\\build_src.py\", line 288, in build_library_sources\n",
      "          sources = self.generate_sources(sources, (lib_name, build_info))\n",
      "                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\command\\build_src.py\", line 378, in generate_sources\n",
      "          source = func(extension, build_dir)\n",
      "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"numpy\\core\\setup.py\", line 671, in get_mathlib_info\n",
      "          st = config_cmd.try_link('int main(void) { return 0;}')\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\command\\config.py\", line 243, in try_link\n",
      "          self._link(body, headers, include_dirs,\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\command\\config.py\", line 162, in _link\n",
      "          return self._wrap_method(old_config._link, lang,\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\command\\config.py\", line 96, in _wrap_method\n",
      "          ret = mth(*((self,)+args))\n",
      "                ^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\command\\config.py\", line 137, in _link\n",
      "          (src, obj) = self._compile(body, headers, include_dirs, lang)\n",
      "                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\command\\config.py\", line 105, in _compile\n",
      "          src, obj = self._wrap_method(old_config._compile, lang,\n",
      "                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\command\\config.py\", line 96, in _wrap_method\n",
      "          ret = mth(*((self,)+args))\n",
      "                ^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\command\\config.py\", line 132, in _compile\n",
      "          self.compiler.compile([src], include_dirs=include_dirs)\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\_msvccompiler.py\", line 401, in compile\n",
      "          self.spawn(args)\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-build-env-w8y7n2x1\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\_msvccompiler.py\", line 505, in spawn\n",
      "          return super().spawn(cmd, env=env)\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "        File \"C:\\Users\\raphe\\AppData\\Local\\Temp\\pip-install-qd52heba\\numpy_ac2299dfbfd744f4ac2325f4433d4939\\numpy\\distutils\\ccompiler.py\", line 90, in <lambda>\n",
      "          m = lambda self, *args, **kw: func(self, *args, **kw)\n",
      "                                        ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "      TypeError: CCompiler_spawn() got an unexpected keyword argument 'env'\n",
      "      [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "error: metadata-generation-failed\n",
      "\n",
      "× Encountered error while generating package metadata.\n",
      "╰─> See above for output.\n",
      "\n",
      "note: This is an issue with the package mentioned above, not pip.\n",
      "hint: See above for details.\n"
     ]
    }
   ],
   "source": [
    "pip install volpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "304b8fad",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'volpy'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mModuleNotFoundError\u001b[39m                       Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[7]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mvolpy\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mv\u001b[39;00m\n",
      "\u001b[31mModuleNotFoundError\u001b[39m: No module named 'volpy'"
     ]
    }
   ],
   "source": [
    "import volpy as v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc1ab7d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10.300022208827253\n"
     ]
    }
   ],
   "source": [
    "print(v.bs_price(100,100,0.02,0.01,1.0,0.25, True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "497bd3fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.5599009112631625,\n",
       " 0.015585304030313534,\n",
       " 38.96326007578383,\n",
       " -5.2243079765595954,\n",
       " 45.69006891748898)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g = v.bs_price_greeks(100,100,0.02,0.01,1.0,0.25, True)\n",
    "g.delta, g.gamma, g.vega, g.theta, g.rho"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5126a0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.2499999999999997, True, 4)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p = v.bs_price(100,100,0.02,0.01,1.0,0.25, True)\n",
    "\n",
    "iv = v.implied_vol(100,100,0.02,0.01,1.0, p, True, 0.5, 1e-8)\n",
    "iv.iv, iv.converged, iv.newton_iters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "091f8206",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10.300022208827253,\n",
       " 10.301819237147956,\n",
       " 0.007777576375011837,\n",
       " 200000,\n",
       " '0.008s')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analytic = v.bs_price(100,100,0.02,0.01,1.0,0.25, True)\n",
    "t0 = time.time()\n",
    "mc = v.mc_euro_gbm(100,100,0.02,0.01,1.0,0.25, True, 200_000, 42)\n",
    "dt = time.time() - t0\n",
    "analytic, mc.price, mc.stderr, mc.paths, f\"{dt:.3f}s\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f6af2ca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def iv7(S,K,r,q,T,price,is_call, v0=0.25, tol=1e-8):\n",
    "    return v.implied_vol(float(S), float(K), float(r), float(q), float(T),\n",
    "                            float(price), bool(is_call), float(v0), float(tol)).iv\n",
    "\n",
    "ivs = [iv7(S, K, r, q, T, m, True) for K, m in zip(Ks, mids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c6c89891",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'np' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[3]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m S, r, q, T, true_vol = \u001b[32m100\u001b[39m, \u001b[32m0.02\u001b[39m, \u001b[32m0.01\u001b[39m, \u001b[32m1.0\u001b[39m, \u001b[32m0.25\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m Ks = \u001b[43mnp\u001b[49m.linspace(\u001b[32m60\u001b[39m, \u001b[32m140\u001b[39m, \u001b[32m17\u001b[39m)\n\u001b[32m      3\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mtrue_smile\u001b[39m(K):\n\u001b[32m      4\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[32m0.25\u001b[39m + \u001b[32m0.0005\u001b[39m * (K - \u001b[32m100\u001b[39m)**\u001b[32m2\u001b[39m / \u001b[32m100\u001b[39m\n",
      "\u001b[31mNameError\u001b[39m: name 'np' is not defined"
     ]
    }
   ],
   "source": [
    "S, r, q, T, true_vol = 100, 0.02, 0.01, 1.0, 0.25\n",
    "Ks = np.linspace(60, 140, 17)\n",
    "def true_smile(K):\n",
    "    return 0.25 + 0.0005 * (K - 100)**2 / 100\n",
    "mids = [v.bs_price(S, K, r, q, T, true_smile(K), True) for K in Ks]\n",
    "ivs  = [v.implied_vol(S, K, r, q, T, m, True, 0.25, 1e-8).iv for K, m in zip(Ks, mids)]\n",
    "\n",
    "plt.plot(Ks/S, ivs, marker='o')\n",
    "plt.xlabel('Moneyness K/S'); plt.ylabel('Implied vol'); plt.title('Smile (synthetic)')\n",
    "plt.grid(True); plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
